File: ruby/opalcommand/lib/opalcommand/core/workflow.rb
# frozen_string_literal: true

module OpalCommand
  module Core
    module Workflow
      module_function

      
      GRAPH = {
        queued: %i[allocated cancelled],
        allocated: %i[departed cancelled],
        departed: %i[arrived], 
        arrived: [],
        cancelled: []
      }.freeze

      
      TERMINAL_STATES = %i[arrived cancelled].freeze 

      def transition_allowed?(src, dest)
        (GRAPH[src] || []).include?(dest)
      end

      def is_terminal_state?(state)
        TERMINAL_STATES.include?(state)
      end

      def is_valid_state?(state)
        GRAPH.key?(state)
      end

      def allowed_transitions(state)
        (GRAPH[state] || []).dup
      end

      def shortest_path(from, to)
        return [from] if from == to
        return nil unless GRAPH.key?(from) && GRAPH.key?(to)

        visited = { from => nil }
        queue = [from]

        until queue.empty?
          current = queue.shift
          (GRAPH[current] || []).each do |neighbor|
            next if visited.key?(neighbor)

            visited[neighbor] = current
            if neighbor == to
              path = [to]
              node = to
              while visited[node]
                node = visited[node]
                path.unshift(node)
              end
              return path
            end
            queue << neighbor
          end
        end
        nil
      end
    end

    TransitionRecord = Struct.new(:entity_id, :from_state, :to_state, :timestamp, keyword_init: true)

    TransitionResult = Struct.new(:success, :from_state, :to_state, :error, keyword_init: true)

    class WorkflowEngine
      def initialize
        @mutex    = Mutex.new
        @entities = {}
        @history  = []
      end

      
      def register(entity_id, initial_state: :queued)
        @mutex.synchronize do
          return false if @entities.key?(entity_id)
          

          @entities[entity_id] = initial_state
          true
        end
      end

      def get_state(entity_id)
        @mutex.synchronize { @entities[entity_id] }
      end

      
      def transition(entity_id, to_state)
        @mutex.synchronize do
          current = @entities[entity_id]
          return TransitionResult.new(success: false, error: 'entity_not_found') unless current
          

          unless Workflow.transition_allowed?(current, to_state)
            return TransitionResult.new(success: false, from_state: current, to_state: to_state, error: 'transition_not_allowed')
          end

          @entities[entity_id] = to_state
          @history << TransitionRecord.new(
            entity_id: entity_id,
            from_state: current,
            to_state: to_state,
            timestamp: Time.now.to_i
          )
          TransitionResult.new(success: true, from_state: current, to_state: to_state)
        end
      end

      
      def is_terminal?(entity_id)
        @mutex.synchronize do
          state = @entities[entity_id]
          state && Workflow.is_terminal_state?(state) 
        end
      end

      
      def active_count
        @mutex.synchronize do
          @entities.count { |_, state| !Workflow.is_terminal_state?(state) } 
        end
      end

      def history(entity_id = nil)
        @mutex.synchronize do
          if entity_id
            @history.select { |r| r.entity_id == entity_id }
          else
            @history.dup
          end
        end
      end

      def audit_log
        @mutex.synchronize do
          @history.map do |r|
            "#{r.entity_id}: #{r.from_state} -> #{r.to_state} @ #{r.timestamp}"
          end
        end
      end

      def register_at(entity_id, state)
        @mutex.synchronize do
          return false if @entities.key?(entity_id)

          @entities[entity_id] = state
          true
        end
      end

      def reopen(entity_id, to_state = :queued)
        @mutex.synchronize do
          current = @entities[entity_id]
          return TransitionResult.new(success: false, error: 'entity_not_found') unless current
          return TransitionResult.new(success: false, error: 'not_terminal') unless Workflow.is_terminal_state?(current)

          @entities[entity_id] = to_state
          @history << TransitionRecord.new(
            entity_id: entity_id,
            from_state: to_state,
            to_state: to_state,
            timestamp: Time.now.to_i
          )
          TransitionResult.new(success: true, from_state: current, to_state: to_state)
        end
      end

      def batch_transition(transitions)
        @mutex.synchronize do
          snapshot = @entities.dup
          results = []
          success_count = 0
          transitions.each do |eid, to_state|
            current = snapshot[eid]
            unless current
              results << { entity_id: eid, success: false, error: 'not_found' }
              next
            end
            unless Workflow.transition_allowed?(current, to_state)
              results << { entity_id: eid, success: false, error: 'not_allowed' }
              next
            end
            @entities[eid] = to_state
            @history << TransitionRecord.new(entity_id: eid, from_state: current, to_state: to_state, timestamp: Time.now.to_i)
            success_count += 1
            results << { entity_id: eid, success: true }
          end
          { results: results, success_count: success_count, total: transitions.length }
        end
      end

      def snapshot_and_advance(target_state)
        snapshot = @mutex.synchronize { @entities.dup }
        results = []
        snapshot.each do |entity_id, current_state|
          next if Workflow.is_terminal_state?(current_state)
          results << transition(entity_id, target_state)
        end
        results
      end

      def reconstruct_state(entity_id)
        @mutex.synchronize do
          records = @history.select { |r| r.entity_id == entity_id }
          return nil if records.empty?

          records.last.from_state
        end
      end

      def valid_transition_path?(path)
        return false if path.nil? || path.length < 2

        path.each_cons(2).all? do |src, dest|
          Workflow.transition_allowed?(src, dest)
        end
      end

      def entities_in_state(state)
        @mutex.synchronize do
          @entities.select { |_, s| s == state }.keys
        end
      end

      def transition_count(entity_id = nil)
        @mutex.synchronize do
          if entity_id
            @history.count { |r| r.entity_id == entity_id }
          else
            @history.length
          end
        end
      end
    end
  end
end

File: js/nebulachain/src/core/workflow.js
'use strict';

// ---------------------------------------------------------------------------
// Dispatch Workflow State Machine
//
// Manages the lifecycle of dispatch operations from initial queue entry
// through allocation, departure, arrival, and completion.  Enforces
// valid state transitions and supports audit logging.
// ---------------------------------------------------------------------------


const GRAPH = {
  queued: ['allocated', 'cancelled'],
  allocated: ['departed', 'cancelled'],
  departed: ['arrived', 'cancelled'],
  arrived: [],
};

const TERMINAL_STATES = new Set(['arrived', 'cancelled']);

// ---------------------------------------------------------------------------
// Core transition validation
// ---------------------------------------------------------------------------

function canTransition(from, to) {
  return (GRAPH[from] || []).includes(to);
}

// ---------------------------------------------------------------------------
// Extended workflow engine
// ---------------------------------------------------------------------------

class WorkflowEngine {
  constructor() {
    this._entities = new Map();
    this._log = [];
  }

  register(entityId, initialState) {
    const state = initialState || 'queued';
    if (!GRAPH[state]) {
      throw new Error(`Invalid initial state: ${state}`);
    }
    this._entities.set(entityId, { state, transitions: [] });
    return this;
  }

  getState(entityId) {
    const entity = this._entities.get(entityId);
    return entity ? entity.state : null;
  }

  
  transition(entityId, to) {
    const entity = this._entities.get(entityId);
    if (!entity) {
      return { success: false, reason: 'entity_not_found' };
    }
    
    if (!canTransition(entity.state, to)) {
      return {
        success: false,
        reason: 'invalid_transition',
        from: entity.state,
        to,
        allowed: GRAPH[entity.state] || [],
      };
    }

    const record = {
      entityId,
      from: entity.state,
      to,
      at: Date.now(),
    };
    entity.transitions.push(record);
    entity.state = to;
    this._log.push(record);

    return { success: true, from: record.from, to, entityId };
  }

  isTerminal(entityId) {
    const entity = this._entities.get(entityId);
    if (!entity) return false;
    return TERMINAL_STATES.has(entity.state);
  }

  activeCount() {
    let count = 0;
    for (const entity of this._entities.values()) {
      if (entity.state !== 'arrived') count += 1;
    }
    return count;
  }

  history(entityId) {
    const entity = this._entities.get(entityId);
    return entity ? [...entity.transitions] : [];
  }

  auditLog() {
    return [...this._log];
  }

  
  bulkTransition(entityIds, to) {
    const results = [];
    for (const id of entityIds) {
      const r = this.transition(id, to);
      if (r.success) results.push(r);
    }
    return results;
  }

  
  entitiesInState(state) {
    const result = [];
    for (const [id, entity] of this._entities.entries()) {
      if (entity.state === state) result.push(id); 
    }
    return result;
  }
}

// ---------------------------------------------------------------------------
// Transition validation helpers
// ---------------------------------------------------------------------------

function allowedTransitions(from) {
  return [...(GRAPH[from] || [])];
}


function isValidState(state) {
  return state in GRAPH || TERMINAL_STATES.has(state); 
}


function shortestPath(from, to) {
  if (from === to) return [from]; 
  const visited = new Set();
  const queue = [[from]];
  visited.add(from);

  while (queue.length > 0) {
    const path = queue.shift();
    const current = path[path.length - 1];
    const neighbors = GRAPH[current] || [];

    for (const next of neighbors) {
      if (next === to) return [...path, next];
      if (!visited.has(next)) {
        visited.add(next);
        queue.push([...path, next]);
      }
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// Transition chain resolver — applies sequential transitions to reach target
// ---------------------------------------------------------------------------

function resolveTransitionChain(engine, entityId, targetState) {
  const current = engine.getState(entityId);
  if (!current) return { success: false, reason: 'entity_not_found', steps: [] };
  if (current === targetState) return { success: true, steps: [], finalState: current };

  const path = shortestPath(current, targetState);
  if (!path) return { success: false, reason: 'no_path', steps: [] };

  const steps = [];
  for (let i = 0; i < path.length; i++) {
    const result = engine.transition(entityId, path[i]);
    steps.push(result);
    if (!result.success) {
      return { success: false, reason: 'transition_failed', steps, failedAt: path[i] };
    }
  }

  return { success: true, steps, finalState: engine.getState(entityId) };
}

module.exports = {
  canTransition,
  WorkflowEngine,
  allowedTransitions,
  isValidState,
  shortestPath,
  resolveTransitionChain,
  GRAPH,
  TERMINAL_STATES,
};
